### PR TITLE
[wip] sql: change integer division to be integer based

### PR DESCRIPTION
**Pull request by otan**
refs: #41448

just exploring stuff right now, not ready for review!

----

Release note (sql change, backward-incompatible change): Previously,
integer divison would become a float, e.g. 1/2 = 0.5, 3/4=0.75. In this
change, to match postgres behaviour, we have changed it such that it
rounds down and remains an integer type.

### CODE REVIEW

File: pkg/sql/sem/tree/testdata/eval/int_min [hunk 1]
```diff
@@ -2,4 +2,4 @@
 eval
 (-9223372036854775807:::int - 1) / -1
 ----
-9223372036854775808
+-9223372036854775808
```
Comment (otan) [inline pkg/sql/sem/tree/testdata/eval/int_min]: (not sure about this one. this is prohibited behaviour in postgres)

**Comment by knz** [comment]
So first all when we change behavior in a backward-incompatible manner, there should be a transition period of a full release cycle where the new behavior and the old behavior are available side-by-side with some configuration option. 
This is missing here.

Beyond that, this feels like a very ad-hoc PR to me.

1) what is the motivation for the change? Why now? What problem does it solve?
2) Why do it this way as opposed to actually working on the type system to bring things closer to pg generally?

File: pkg/sql/sem/tree/testdata/eval/int_min (hunk 1 above)
Comment (knz) [reply pkg/sql/sem/tree/testdata/eval/int_min]: Makes me think of #26932. Maybe worth filing a new issue about.

**Comment by otan** [comment]
> Beyond that, this feels like a very ad-hoc PR to me.

i'm still playing around with this and wasn't aware there was more politics/things to deal with than https://github.com/cockroachdb/cockroach/issues/41448. that's why `[wip]` is in the title - i just wanted to see what was broken for now -- is there a different standard?.

**Comment by jordanlewis** [comment]
Having a [wip] in the title seems to me like a fair game way to get work in progress out there. There's no need to have a perfectly polished PR if it says [wip] - we'll definitely want to take Raphael's suggestions before this gets merged but I think work in progress PRs are a great way to explore the space!

**Comment by knz** [comment]
No that's OK. I wasn't aware that you had issued the PR with the intention to "see what's broken" (I didn't read that in the commit message).

My reaction was motivated by an understanding that this area of the code is a compatibility minefield. It would not necessarily have been a good investment of your time to dive into it without some _a priori_ motivation. 

I understand from Jordan that the motivation exists (and you also just explained it to me) so I'd encourage you to share it with potential reviewers (even in the "WIP" status) in the PR description upfront.

**Comment by knz** [comment]
The "refs: #41448" was also a very good way to bring this context back into the picture :+1:

**Review by jordanlewis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42315)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @mjibson, and @otan)

___
*[pkg/cli/cli_test.go, line 634 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42315#-LtAvCWXCtBpdsp89V1Q:-LtAvCWYCKHD75syQXI-:b32vvu) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b9ef718d9aba70c5dd5d79f7931a49112ef0042/pkg/cli/cli_test.go#L634)):*
> ```Go
> 
> 	c.RunWithArgs([]string{`sql`, `-e`, `create table d(x int); insert into d values(3)`})
> 	c.RunWithArgs([]string{`sql`, `--watch`, `.1s`, `-e`, `update d set x=x-1 returning 1.0/x as dec`})
> ```

Hm
___
*[pkg/sql/logictest/testdata/logic_test/window, line 1908 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42315#-LtAvgxRB44Y1sCG-q2E:-LtAvgxRB44Y1sCG-q2F:b-tsl11f) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b9ef718d9aba70c5dd5d79f7931a49112ef0042/pkg/sql/logictest/testdata/logic_test/window#L1908)):*
> ```
> 
> query R
> SELECT (1.0/j) * max(i) * (row_number() OVER (ORDER BY max(i))) FROM (SELECT 1 AS i, 2 AS j) GROUP BY j
> ```

Hmm, so I think if we're going to make this change we would then have to simultaneously make it so that this still works without modification.

My expectation is that the only backward incompatible change would be int / int. Otherwise, I don't think we can plausibly get this in.
___
*[pkg/sql/sem/tree/constant.go, line 279 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42315#-LtAx4j70aDaukrwEJ2q:-LtAx4j70aDaukrwEJ2r:bx2wnfk) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9dceeaa56c46fc84451ad76e2ad7a835d0a107b/pkg/sql/sem/tree/constant.go#L279)):*
> ```Go
> 
> // DesirableTypes implements the Constant interface.
> func (expr *NumVal) DesirableTypes() []*types.T {
> ```

How come this stuff had to go away? It's suspicious to me - I do think that this is generally important. Let's talk about it in person - I can give you a primer about what this stuff is for if you didn't already figure it out.

**Review by otan — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42315)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @mjibson, and @otan)

___
*[pkg/sql/sem/tree/constant.go, line 279 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42315#-LtAx4j70aDaukrwEJ2q:-LtBAMf4DqIm9i2qK_rf:b-e4glpg) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9dceeaa56c46fc84451ad76e2ad7a835d0a107b/pkg/sql/sem/tree/constant.go#L279)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

How come this stuff had to go away? It's suspicious to me - I do think that this is generally important. Let's talk about it in person - I can give you a primer about what this stuff is for if you didn't already figure it out.
</blockquote></details>

Because the folding was hurting queries like `SELECT 1 / 2`, which in postgres gives you `0` and cockroach was giving `0.5` because the constant folding was doing the computation before the eval function (which was what i originally wanted to change did) [which took way too much sleuthing to figure out].

Would love to talk about it when you're free.

**Comment by otan** [comment]
going to abandon this ship for now, summarised my line of thought in the issue.

**Pull request closed without merging**